From 8b70fc71c98d4ec95319797c09f72b9c9b0f426e Mon Sep 17 00:00:00 2001
From: Igor Grinberg <grinberg@compulab.co.il>
Date: Sat, 2 Mar 2013 00:24:18 +0200
Subject: ARM: i.MX6: cm-fx6: adobt the display init code

1) Refactor the display initialization code to be cm-fx6 specific.
2) Add support for DataImage SCF0403 LCD panel.

Signed-off-by: Igor Grinberg <grinberg@compulab.co.il>
---
 arch/arm/mach-mx6/board-cm-fx6.c |  130 +++++++++++++++++++++-----------------
 1 file changed, 71 insertions(+), 59 deletions(-)

diff --git a/arch/arm/mach-mx6/board-cm-fx6.c b/arch/arm/mach-mx6/board-cm-fx6.c
index 75d5575..2e08cf9 100644
--- a/arch/arm/mach-mx6/board-cm-fx6.c
+++ b/arch/arm/mach-mx6/board-cm-fx6.c
@@ -706,29 +706,7 @@ static struct imx_asrc_platform_data imx_asrc_data = {
 	.clk_map_ver	= 2,
 };
 
-static struct ipuv3_fb_platform_data sabr_fb_data[] = {
-	{ /*fb0*/
-	.disp_dev		= "ldb",
-	.interface_pix_fmt	= IPU_PIX_FMT_RGB666,
-	.mode_str		= "LDB-XGA",
-	.default_bpp		= 16,
-	.int_clk		= false,
-	}, {
-	.disp_dev		= "ldb",
-	.interface_pix_fmt	= IPU_PIX_FMT_RGB666,
-	.mode_str		= "LDB-XGA",
-	.default_bpp		= 16,
-	.int_clk		= false,
-	}, {
-	.disp_dev		= "lcd",
-	.interface_pix_fmt	= IPU_PIX_FMT_RGB565,
-	.mode_str		= "CLAA-WVGA",
-	.default_bpp		= 16,
-	.int_clk		= false,
-	}
-};
-
-static void hdmi_init(int ipu_id, int disp_id)
+static void cm_fx6_hdmi_init(int ipu_id, int disp_id)
 {
 	int hdmi_mux_setting;
 	int max_ipu_id = cpu_is_mx6q() ? 1 : 0;
@@ -778,30 +756,47 @@ static void hdmi_disable_ddc_pin(void)
 			ARRAY_SIZE(cm_fx6_q_i2c2_pads));
 }
 
-static struct fsl_mxc_hdmi_platform_data hdmi_data = {
-	.init = hdmi_init,
+static struct fsl_mxc_hdmi_platform_data cm_fx6_hdmi_data = {
+	.init = cm_fx6_hdmi_init,
 	.enable_pins = hdmi_enable_ddc_pin,
 	.disable_pins = hdmi_disable_ddc_pin,
 };
 
-static struct fsl_mxc_hdmi_core_platform_data hdmi_core_data = {
+static struct fsl_mxc_hdmi_core_platform_data cm_fx6_hdmi_core_data = {
 	.ipu_id		= 0,
 	.disp_id	= 0,
 };
 
-static struct fsl_mxc_lcd_platform_data lcdif_data = {
-	.ipu_id		= 0,
-	.disp_id	= 0,
-	.default_ifmt	= IPU_PIX_FMT_RGB565,
+static void __init cm_fx6_init_hdmi(void)
+{
+	struct platform_device * pdev;
+
+	pdev = imx6q_add_mxc_hdmi_core(&cm_fx6_hdmi_core_data);
+	if (IS_ERR(pdev)) {
+		pr_err("%s: HDMI core register failed: %ld\n",
+		       __func__, PTR_ERR(pdev));
+		return;
+	}
+
+	pdev = imx6q_add_mxc_hdmi(&cm_fx6_hdmi_data);
+	if (IS_ERR(pdev))
+		pr_err("%s: HDMI register failed: %ld\n",
+		       __func__, PTR_ERR(pdev));
+}
+
+static struct ipuv3_fb_platform_data cm_fx6_fb_data[] = {
+	{
+		.disp_dev		= "lcd",
+		.interface_pix_fmt	= IPU_PIX_FMT_RGB666,
+		.mode_str		= "SCF04-WVGA",
+		.int_clk		= false,
+	}
 };
 
-static struct fsl_mxc_ldb_platform_data ldb_data = {
-	.ipu_id		= 1,
+static struct fsl_mxc_lcd_platform_data cm_fx6_lcdif_data = {
+	.ipu_id		= 0,
 	.disp_id	= 0,
-	.ext_ref	= 1,
-	.mode		= LDB_SEP0,
-	.sec_ipu_id	= 0,
-	.sec_disp_id	= 1,
+	.default_ifmt	= IPU_PIX_FMT_RGB666,
 };
 
 static struct imx_ipuv3_platform_data ipu_data[] = {
@@ -814,6 +809,45 @@ static struct imx_ipuv3_platform_data ipu_data[] = {
 	},
 };
 
+static void __init cm_fx6_init_display(void)
+{
+	int i;
+	struct platform_device * pdev;
+
+	cm_fx6_init_hdmi();
+
+	pdev = imx6q_add_ipuv3(0, &ipu_data[0]);
+	if (IS_ERR(pdev)) {
+		pr_err("%s: IPU 0 register failed: %ld\n",
+		       __func__, PTR_ERR(pdev));
+		return;
+	}
+
+	if (cpu_is_mx6q()) {
+		pdev = imx6q_add_ipuv3(1, &ipu_data[1]);
+		if (IS_ERR(pdev))
+			pr_err("%s: IPU 1 register failed: %ld\n",
+			       __func__, PTR_ERR(pdev));
+	}
+
+	for (i = 0; i < ARRAY_SIZE(cm_fx6_fb_data); i++) {
+		pdev = imx6q_add_ipuv3fb(i, &cm_fx6_fb_data[i]);
+		if (IS_ERR(pdev))
+			pr_err("%s: fb%d register failed: %ld\n",
+			       __func__, i, PTR_ERR(pdev));
+	}
+
+	pdev = imx6q_add_vdoa();
+	if (IS_ERR(pdev))
+		pr_err("%s: VDOA register failed: %ld\n",
+		       __func__, PTR_ERR(pdev));
+
+	pdev = imx6q_add_lcdif(&cm_fx6_lcdif_data);
+	if (IS_ERR(pdev))
+		pr_err("%s: lcd interface register failed: %ld\n",
+		       __func__, PTR_ERR(pdev));
+}
+
 #if defined(CONFIG_BACKLIGHT_PWM) || defined(CONFIG_BACKLIGHT_PWM_MODULE)
 static struct platform_pwm_backlight_data sb_fx6_pwm_backlight_data = {
 	.pwm_id		= 2,
@@ -1034,7 +1068,6 @@ static inline void cm_fx6_init_led(void) {}
  */
 static void __init cm_fx6_init(void)
 {
-	int i;
 	int ret;
 
 	iomux_v3_cfg_t *common_pads = NULL;
@@ -1099,26 +1132,7 @@ static void __init cm_fx6_init(void)
 	pu_reg_id = arm2_dvfscore_data.pu_id;
 	cm_fx6_init_uart();
 
-	imx6q_add_mxc_hdmi_core(&hdmi_core_data);
-
-	imx6q_add_ipuv3(0, &ipu_data[0]);
-	if (cpu_is_mx6q())
-		imx6q_add_ipuv3(1, &ipu_data[1]);
-
-	if (cpu_is_mx6dl()) {
-		ldb_data.ipu_id = 0;
-		ldb_data.disp_id = 0;
-		for (i = 0; i < ARRAY_SIZE(sabr_fb_data) / 2; i++)
-			imx6q_add_ipuv3fb(i, &sabr_fb_data[i]);
-	} else {
-		for (i = 0; i < ARRAY_SIZE(sabr_fb_data); i++)
-			imx6q_add_ipuv3fb(i, &sabr_fb_data[i]);
-	}
-
-	imx6q_add_vdoa();
-	imx6q_add_lcdif(&lcdif_data);
-	imx6q_add_ldb(&ldb_data);
-	imx6q_add_v4l2_output(0);
+	cm_fx6_init_display();
 
 	imx6q_add_imx_snvs_rtc();
 
@@ -1128,8 +1142,6 @@ static void __init cm_fx6_init(void)
 	cm_fx6_init_led();
 	cm_fx6_spi_init();
 
-	imx6q_add_mxc_hdmi(&hdmi_data);
-
 	imx6q_add_anatop_thermal_imx(1, &cm_fx6_anatop_thermal_data);
 
 	imx6_init_fec(cm_fx6_fec_data);
-- 
1.7.9.5

